Avoid maps. Use vertex neighbors for sort.

gassmoeller opened the pull request:
In a similar fashion as #1258 this PR tries to optimize particle runtime in adaptive parallel models. It includes three improvements:
- `send_recv_particles` no longer takes a multimap, but rather a vector of vectors of particles to send. This should have been the behaviour from the beginning, because there is no reason to keep them in a multimap here (we loop over them particle-by-particle, not cell-by-cell, and we excactly know how many neighbors we have), and the vector can be filled faster. This results in a number of container changes in other functions as well, but replaces another case of `NlogN` complexity by linear complexity.
- The `sort_particles_in_subdomains_and_cells` function had a major scalability flaw by only checking face neighbors of the current cell before doing a global search (in the local domain) for the cells of the particles. This missed all vertex_neighbors (and edge_neighbors in 3D), and created a quadratic complexity (N_cells times N_particles) that becomes dominant for more than 50,000 particles per process in box models (larger numbers for spherical models). This was not visible before the other particle optimizations, because the linear work performed there was more expensive.
- The order in which neighbor cells are searched now no longer depends on a distance (between particle and face center), but on a direction (more precisely: the scalar product / angle between the vector 'closest-vertex to cell center' and the vector 'closest-vertex to particle'). This avoids problems where one neighbor is refined, and therefore all distances will be smaller than for the not refined neighbors. Now we will most likely find each particle in the first neighbor cell we check, unless the cells are strongly distorted.

All in all a 3D box model with 15 million particles on 4 processes the `sort_particles_in_subdomains_and_cells` function is now 4 times faster (1.8% instead of 7.8% of total runtime), and the container changes also sped up the `exchange_ghosts` function by 25% (8.8% instead of 11% of total runtime).

bangerth commented on source/particle/world.cc:
```diff
@@ -35,6 +35,24 @@ namespace aspect
+    {
+      template <int dim>
+      unsigned int
+      compare_particle_association(const std::pair<typename Triangulation<dim>::active_cell_iterator, Tensor<1,dim> > &cell_a,
```
I think this should somehow say that we compare two cells for how far their centers are in a given direction specified by the third argument. Out of curiosity, why not compare their centers for *distance* to the given target point where we expect the particle to land? I think we talked about this last week, but I don't recall why that would not be a good idea. Currently, a neighboring cell whose center is aligned with the direction but that is far too far in this direction will win this sort operation.

bangerth replied:
Oh, or are the second parts of the pairs also normalized direction vectors?

bangerth commented on source/particle/world.cc:
```diff
@@ -833,21 +924,31 @@ namespace aspect
+             ExcMessage("The particles to send to other processes should be sorted into a vector "
+                        "containing as many vectors of particles as there are neighbor processes. This "
+                        "is not the case for an unknown reason. Contact the developers if you encounter "
+                        "this error."));
```
Not wrong. We commonly use `ExcInternalError()` for this, but not worth changing now.

bangerth commented on source/particle/world.cc:
```diff
@@ -551,6 +569,21 @@ namespace aspect
+    World<dim>::get_subdomain_id_to_neighbor_map() const
+    {
+      std::map<unsigned int, unsigned int> subdomain_id_to_neighbor_map;
+      const std::set<unsigned int> ghost_owners = this->get_triangulation().ghost_owners();
```
See https://github.com/dealii/dealii/pull/3300

gassmoeller commented:
I think I addressed all comments, but I had to change the structure quite a bit (forgot to account for  hanging nodes) so better give it another look. 
About the direction vs. distance sorting: Imagine a situation in which a particle wanders into a neighbor cell of the same level, but close to a another refined neighbor cell. A comparison of the distances between closest vertex and particle and closest vertex and cell center will favor the refined cell (because all points will be closer). Comparing the direction will give a better criterion which cell to check first.

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -407,6 +407,15 @@ namespace aspect
+         * Get a map between subdomain id and the neighbor index. In other words
+         * the returned map answers the question: Given a subdomain id, which
+         * neighbor (in terms of a contiguous number from 0 to n_neighbors)
+         * owns this subdomain?
```
Would you mind changing this to say "which neighbor of the current processor's subdomain (in terms of ...)" to make clear that this is information that will be different on every processor.

This is a function that I think we ought to move to deal.II in the long term. But that can happen after this patch.

bangerth commented on source/particle/world.cc:
```diff
@@ -35,6 +35,57 @@ namespace aspect
+      compare_particle_association(const unsigned int a,
+                                   const unsigned int b,
+                                   const Tensor<1,dim> &particle_direction,
+                                   const std::vector<Tensor<1,dim> > &center_directions)
```
I assume the vectors in the last two arguments are normalized to length 1? If so maybe add that this function assumes so.
